✨ Add command and client for `clusterctl alpha rollout restart`

Arvinderpal opened the pull request:
**What this PR does / why we need it**:
Adds command and client for the `clusterctl alpha rollout restart`.
Also implements `restart` sub-command for MachineDeployments. 

Fixes:
https://github.com/kubernetes-sigs/cluster-api/issues/3401

Related Issue:
https://github.com/kubernetes-sigs/cluster-api/issues/3439

Arvinderpal commented:
Example restart command:
`clusterctl alpha rollout restart --namespace default machinedeployment/test-md-0`

fabriziopandini commented:
/ok-to-test

Arvinderpal commented on cmd/clusterctl/client/client.go:
```diff
@@ -65,6 +65,15 @@ type Client interface {
 	ProcessYAML(options ProcessYAMLOptions) (YamlPrinter, error)
+
+	// Interface for alpha features in clusterctl
+	AlphaClient
```
@joelanford @alexeldeib Can you help me make sense of why this is flagged as an "incompatible change" with `go-apidiff`? I also tried removing the embedded interface and just putting the func directly in Client interface, but similar issue. 
```
*** Running go-apidiff ***

sigs.k8s.io/cluster-api/cmd/clusterctl/client
  Incompatible changes:
  - AlphaClient.RolloutRestart: added
  Compatible changes:
  - (*clusterctlClient).RolloutRestart: added
  - AlphaClient: added
  - RolloutRestartOptions: added

sigs.k8s.io/cluster-api/cmd/clusterctl/client/rollout
  Compatible changes:
  - ObjectRestarter: added

sigs.k8s.io/cluster-api/cmd/clusterctl/cmd/rollout
  Compatible changes:
  - NewCmdRolloutRestart: added
+ EXIT_VALUE=1
+ set +o xtrace
```
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/3838/pull-cluster-api-apidiff-main/1319649725397340160/build-log.txt

alexeldeib replied:
The test is more advisory than required -- it's okay that it fails for PRs to be merged some times.  The PR does make all of the changes mentioned, but they all seem reasonable to me (actually a bit surprised the first one is incompatible and the second one is compatible, would have expected those to be flipped since adding to an existing interface would require implementers to update their code).

Arvinderpal replied:
Ah okay. Unless others see any problems with what I'm doing, I'll just ignore this for now.

Arvinderpal commented:
@fabriziopandini Can I get a review of the overall structure of this code? This would unblock me to work on the other sub-commands for rollout (i.e. status/undo/history/pause). Much appreciated.

wfernandes commented on cmd/clusterctl/client/client.go:
```diff
@@ -65,6 +65,15 @@ type Client interface {
+// AlphaClient exposes the alpha features in clusterctl high-level client library.
+type AlphaClient interface {
+	// RolloutRestart provides rollout restart of cluster-api resources
+	RolloutRestart(options RolloutRestartOptions) error
```
I can see that as other "alpha" commands get added in, this interface could grow large. Since it seems that there are many rollout related functions, we can perhaps embed or add a `Rollout rollout.Client` interface here.

Something similar to `cmd/clusterctl/client/repository/client.go` and `cmd/clusterctl/client/cluster/client.go`.

This way all the rollout related behavior can be kept in that interface and that can be just referenced here.

Arvinderpal replied:
I refactored the code. PTAL

Arvinderpal replied:
I'm not sure how that would work? Rollout looks like this:
```
type Rollout interface {
	ObjectRestarter(cluster.Proxy, util.ResourceTuple, string) error
}
```
Where would the `RolloutRestart(options RolloutRestartOptions) error` reside? 

I believe I had addressed @wfernandes comments with a refactor earlier.

wfernandes commented on cmd/clusterctl/internal/util/resource_tuples_test.go:
```diff
@@ -0,0 +1,86 @@
+	}{
+		{
+			name: "valid",
+			args: []string{"machinedeployment/foo"},
```
What happens if we just specify `machinedeployment`? What is the expected behavior?
That is, if the user does `clusterctl alpha rollout restart machinedeployment`.

Arvinderpal replied:
Added a new test case.

wfernandes replied:
So I meant what if we just specified `machinedeployment` without the slash.

If we remove the `/` on line 61 below, we don't get an error but do we expect any `ResourceTuple`s?

That is, if we run `clusterctl alpha rollout restart machinedeployment` are we expecting all the machinedeployments in the specified namespace to rollout or do we error asking the user to specify a name for the machinedeployment?

Arvinderpal replied:
I see your point. Yes, we should throw an error to the user.

wfernandes commented on cmd/clusterctl/client/rollout.go:
```diff
@@ -0,0 +1,79 @@
+	Namespace string
+}
+
+func (c *clusterctlClient) RolloutRestart(options RolloutRestartOptions) error {
```
It would be nice to have some tests for this behavior. WDYT?

Arvinderpal replied:
I added the tests. PTAL

wfernandes commented on cmd/clusterctl/cmd/rollout.go:
```diff
@@ -0,0 +1,56 @@
+	rolloutValidResources = dedent.Dedent(`
+		Valid resource types include:
+
+		   * machinedeployments
```
Since the key we are using is `machinedeployment/foo`, might be nice to keep it consistent.
```suggestion
		   * machinedeployment
```

wfernandes commented on cmd/clusterctl/cmd/rollout.go:
```diff
@@ -0,0 +1,56 @@
+package cmd
+
+import (
+	"github.com/lithammer/dedent"
```
nit: Do we need really need this dependency?

Arvinderpal replied:
removed it.

wfernandes commented on cmd/clusterctl/cmd/rollout/restart.go:
```diff
@@ -0,0 +1,85 @@
+limitations under the License.
+*/
+
+package rollout
```
I noticed that in the `clusterctl/cmd` directory, we have the following additions:
```shell
# Deleted some lines to reduce verbosity.
├── cmd
│   ├── alpha.go
│   ├── config.go
│   ├── config_cluster.go
│   ├── config_provider.go
│   ├── config_repositories.go
│   ├── config_repositories_test.go
│   ├── generate.go
│   ├── generate_yaml.go
│   ├── generate_yaml_test.go
│   ├── get.go
│   ├── get_kubeconfig.go
│   ├── rollout
│   │   └── restart.go
│   ├── rollout.go
│   ├── upgrade.go
│   ├── upgrade_apply.go
│   ├── upgrade_plan.go
...
```

Regarding structure, we have been naming subcommands as defined by `command_subcommand.go`. It seems that if we strictly follow the convention we will have `alpha_rollout_restart.go`. I can see that this may be contentious in the future if we wanted to "graduate" the command from an alpha command into a standard command. So as a compromise, I suggest doing something like this.
```shell
# Deleted some lines to reduce verbosity.
├── cmd
│   ├── alpha.go
│   ├── rollout.go
│   ├── rollout_restart.go
│   ├── rollout_status.go
│   ├── rollout_undo.go
...
```
This way in the future, if we need to "graduate" the rollout cmd from alpha, it will just involve removing it as a subcommand from the `alpha.go` file and no other changes.
WDYT?

Arvinderpal replied:
I had it that way initially, but didn't like the explosion in files in that `/cmd` that this would lead to. So, decided to create a separate directory for each sub-command of `rollout`.

wfernandes replied:
My only concern here is code consistency. I don't necessarily mind that we have these commands in a subdirectory, it's just that we now have some commands in a subdirectory vs the other in `cmd_subcmd.go` format.

Again, not a major concern but maybe @fabriziopandini can weigh in.

wfernandes commented on cmd/clusterctl/client/rollout.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+func normalizeResources(input []string) []string {
+	normalized := []string{}
```
```suggestion
	normalized := make([]string, 0, len(input))
```

wfernandes reviewed (commented):
@Arvinderpal I took a look at the overall structure of the code. A couple of comments.

- I noticed that the issue referenced was for supporting a `rolloutAfter`. Maybe we need a new issue to describe this functionality or rename/retitle issue #3401 to better describe the work being done in this PR.

- Are we planning on having separate PRs for each rollout subcommand - status/undo/history/pause or just one for all of them? I'm personally leaning towards adding each subcommand as a separate PR each one adding new behavior. WDYT?

- Also, we'll need added and updated documentation for each command to the CAPI book.

Arvinderpal commented:
@wfernandes Thank you for the feedback. Much appreciated :smiley: 

> * I noticed that the issue referenced was for supporting a `rolloutAfter`. Maybe we need a new issue to describe this functionality or rename/retitle issue #3401 to better describe the work being done in this PR.

I have updated that particular issue with a new title. 

> * Are we planning on having separate PRs for each rollout subcommand - status/undo/history/pause or just one for all of them? I'm personally leaning towards adding each subcommand as a separate PR each one adding new behavior. WDYT?

Separate would be best. This is already getting rather large and it's only the restart command + skeleton. 

> * Also, we'll need added and updated documentation for each command to the CAPI book.

Yes, I can certainly do that once we have this code ready to merge. I will add that to the tasks in the tracking issue.

wfernandes commented on cmd/clusterctl/client/alpha/rollout.go:
```diff
@@ -0,0 +1,38 @@
+type rollout struct{}
+
+func newRolloutClient(options ...Option) Rollout {
+	client := &rollout{}
+	return client
```
```suggestion
	return &rollout{}
```

wfernandes commented on cmd/clusterctl/client/alpha/rollout.go:
```diff
@@ -0,0 +1,38 @@
+	ObjectRestarter(cluster.Proxy, util.ResourceTuple, string) error
+}
+
+// var _ Rollout = &test.FakeRollout{}
```
remove this comment

wfernandes commented on cmd/clusterctl/client/alpha/rollout_restarter.go:
```diff
@@ -0,0 +1,100 @@
+	}
+	if err := c.Get(context.TODO(), mdObjKey, mdObj); err != nil {
+		return errors.Wrapf(err, "error reading %q %s/%s",
+			mdObj.GroupVersionKind(), mdObj.GetNamespace(), mdObj.GetName())
```
```suggestion
			mdObj.GroupVersionKind(), mdObjKey.GetNamespace(), mdObjKey.GetName())
```

wfernandes commented on cmd/clusterctl/client/alpha/rollout_restarter.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+// getMachineDeployment retrieves the MachineDeployment object corresponding to the name and namespace specified.
+func getMachineDeployment(proxy cluster.Proxy, name, namespace string, mdObj *clusterv1.MachineDeployment) error {
```
Just curious, is there a reason we want to pass in a pointer to the MachineDeployment or can we just return the MachineDeployment pointer from this function. Like
```golang
func getMachineDeployment(proxy cluster.Proxy, name, namespace string) (*clusterv1.MachineDeployment, error)
```

wfernandes commented on cmd/clusterctl/client/alpha/rollout_restarter.go:
```diff
@@ -0,0 +1,100 @@
+
+var validResourceTypes = []string{"machinedeployment"}
+
+func (r *rollout) ObjectRestarter(proxy cluster.Proxy, tuple util.ResourceTuple, namespace string) error {
```
nit: add godoc

wfernandes commented on cmd/clusterctl/client/alpha/rollout_restarter.go:
```diff
@@ -0,0 +1,100 @@
+
+// setRestartedAtAnnotation sets the restartedAt annotation in the MachineDeployment's spec.template.objectmeta.
+func setRestartedAtAnnotation(proxy cluster.Proxy, name, namespace string) error {
+	patch := client.RawPatch(types.MergePatchType, []byte(fmt.Sprintf("{\"spec\":{\"template\":{\"metadata\":{\"annotations\":{\"cluster.x-k8s.io/restartedAt\":\"%v\"}}}}}", time.Now().Format(time.RFC3339))))
```
Suggestion: We've been using `util/patch` in cluster-api to patch objects. See usage example in `controllers/machinedeployment_controller.go`.

wfernandes replied:
Ahh...ok. Didn't notice that.

wfernandes replied:
Regarding this line, could we have a test to assert that the expected annotation on the template of the machine deployment is set?

Arvinderpal replied:
Yes, I added it: Test_ObjectRestarter

wfernandes reviewed (commented):
thanks for making the changes. I just did another quick pass.

Arvinderpal reviewed (commented):
@wfernandes I have incorporated all of your feedback. PTAL

wfernandes commented on cmd/clusterctl/client/alpha/rollout_restarter.go:
```diff
@@ -0,0 +1,105 @@
+		if mdObj == nil {
+			return errors.Errorf("failed to fetch %v/%v", tuple.Resource, tuple.Name)
+		}
+		if mdObj.Spec.Paused {
```
We may want to include a test for this specific case. WDYT?

Arvinderpal replied:
Yes, I added a test: Test_ObjectRestarter

wfernandes reviewed (commented):
@Arvinderpal Thanks for the updates and adding the tests! We are almost there. Just a few comments after I did a manual test against a CAPA cluster and it seemed to work great.

Arvinderpal reviewed (commented):
@wfernandes I addressed all your comments. PTAL again.

wfernandes commented on cmd/clusterctl/client/alpha/rollout_restarter.go:
```diff
@@ -0,0 +1,101 @@
+	switch tuple.Resource {
+	case "machinedeployment":
+		mdObj, err := getMachineDeployment(proxy, tuple.Name, namespace)
+		if err != nil {
+			return errors.Wrapf(err, "failed to fetch %v/%v", tuple.Resource, tuple.Name)
+		}
+		if mdObj == nil {
+			return errors.Errorf("failed to fetch %v/%v", tuple.Resource, tuple.Name)
+		}
```
Apologies for not noticing this earlier. But this is a minor refactor.

```suggestion
		if err != nil || mdObj == nil {
			return errors.Wrapf(err, "failed to fetch %v/%v", tuple.Resource, tuple.Name)
		}
```

wfernandes commented:
@Arvinderpal Thanks for this PR! This LGTM for the most part. 
I noticed that some of the CI jobs were failing. I suggest running `make lint test` or perhaps `./scripts/ci-make.sh`.

@CecileRobertMichon @fabriziopandini Maybe someone can have another look when y'all get the chance.

Arvinderpal commented:
> @Arvinderpal Thanks for this PR! This LGTM for the most part.
> I noticed that some of the CI jobs were failing. I suggest running `make lint test` or perhaps `./scripts/ci-make.sh`.
> 
> @CecileRobertMichon @fabriziopandini Maybe someone can have another look when y'all get the chance.

@wfernandes Thanks. I have cleaned up the CI failures. 
@fabriziopandini @CecileRobertMichon PTAL. 
I will follow up with additional PRs for the other sub-commands.

wfernandes commented:
/lgtm

vincepri commented:
/milestone v0.4.0

wfernandes commented on cmd/clusterctl/cmd/rollout.go:
```diff
@@ -0,0 +1,53 @@
+		`)
+
+	rolloutExample = Examples(`
+		# Force an immediate rollout of machinedeployment
+		clusterctl alpha rollout restart machinedeployment/my-md-0
+
+		# Rollback to the previous revision of a machinedeployment
+		clusterctl alpha rollout undo machinedeployment/my-md-0
+
+		# Check the rollout status of a machinedeployment
+		clusterctl alpha rollout status machinedeployment/my-md-0`)
```
I believe in the issue there was a discussion point about rolling out the control plane as well.

Arvinderpal replied:
The goal is to support kubadmcontrolplane. I had not given much thought to machinepools.

Arvinderpal commented on cmd/clusterctl/client/alpha/rollout.go:
```diff
@@ -0,0 +1,35 @@
+
+// Rollout defines the behavior of a rollout implementation.
+type Rollout interface {
+	ObjectRestarter(cluster.Proxy, util.ResourceTuple, string) error
```
I'm using the kubectl naming conventions. There will be other funcs for status, history, pause. 
Is the context a must? The internal function does not currently need/use it.

Arvinderpal commented on cmd/clusterctl/client/alpha/rollout_restarter.go:
```diff
@@ -0,0 +1,97 @@
+// ObjectRestarter will issue a restart on the specified cluster-api resource.
+func (r *rollout) ObjectRestarter(proxy cluster.Proxy, tuple util.ResourceTuple, namespace string) error {
+	switch tuple.Resource {
+	case "machinedeployment":
```
There will *hopefully* be support for kubeadmcontrolplane soon. We will have a case statement for that as well. 
Just following kubectl convention in terms of code structure. They also make use of switch statements for different objecs like Deployment, StatefulSet, ...

Arvinderpal commented on cmd/clusterctl/client/client.go:
```diff
@@ -160,6 +171,11 @@ func newClusterctlClient(path string, options ...Option) (*clusterctlClient, err
 		client.clusterClientFactory = defaultClusterFactory(client.configClient)
 	}
 
+	if client.alphaClient == nil {
```
done

Arvinderpal commented on cmd/clusterctl/client/alpha/rollout_restarter.go:
```diff
@@ -0,0 +1,97 @@
+func (r *rollout) ObjectRestarter(proxy cluster.Proxy, tuple util.ResourceTuple, namespace string) error {
+	switch tuple.Resource {
+	case "machinedeployment":
+		mdObj, err := getMachineDeployment(proxy, tuple.Name, namespace)
```
done

Arvinderpal reviewed (commented):
@vincepri Thanks for the feedback.

Arvinderpal commented:
@vincepri Can I get another review of this PR? 
If you're good with it, perhaps we can finally merge it. :smile: 
Much appreciated.

Arvinderpal commented:
/assign @vincepri

wfernandes commented:
I can take another look at this sometime today.
/assign @wfernandes

vincepri commented:
LGTM
@wfernandes @fabriziopandini over to you

wfernandes commented on cmd/clusterctl/cmd/rollout.go:
```diff
@@ -0,0 +1,53 @@
+		# Force an immediate rollout of machinedeployment
+		clusterctl alpha rollout restart machinedeployment/my-md-0
+
+		# Rollback to the previous revision of a machinedeployment
+		clusterctl alpha rollout undo machinedeployment/my-md-0
+
+		# Check the rollout status of a machinedeployment
+		clusterctl alpha rollout status machinedeployment/my-md-0`)
```
nit: We can remove these since we aren't supporting them yet. When we do add the subsequent `undo` and `status` subcommands, then we can include these examples.

Arvinderpal replied:
done

wfernandes reviewed (commented):
Just one nit. We should also add some documentation for this command in the CAPI book. See https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/clusterctl/overview.md and 
https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/clusterctl/commands/commands.md

Arvinderpal commented:
> Just one nit. We should also add some documentation for this command in the CAPI book. See https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/clusterctl/overview.md and
> https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/clusterctl/commands/commands.md

If you don't mind, can I update the docs in a followup PR?
I have created a task for this in the tracking issue: https://github.com/kubernetes-sigs/cluster-api/issues/3439

Also, I spoke with @fabriziopandini on slack. He didn't feel the need to review as long as you and @vincepri had done so. 

If the tests pass, I think we should go ahead and merge.

wfernandes commented:
/lgtm

wfernandes commented:
/test pull-cluster-api-test-main

wfernandes commented on cmd/clusterctl/client/rollout_test.go:
```diff
@@ -0,0 +1,166 @@
+	md1 := &clusterv1.MachineDeployment{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "MachineDeployment",
+			APIVersion: "cluster.x-k8s.io/v1alpha3",
```
From the test failures, you may need to update these to v1alpha4. However, I didn't think we needed the APIVersion specified, so I think you can remove this.

wfernandes commented:
@Arvinderpal  Ok..I was able to reproduce this failure by rebasing this branch with the main branch. I think the import in the rollout_test should be `sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha4` but I haven't confirmed that fix yet.

wfernandes commented:
@Arvinderpal Yeah..I think that's it. We recently bumped to using v1alpha4 types everywhere instead of v1alpha3. Please change all the `sigs.k8s.io/cluster-api/api/v1alpha3` imports to `sigs.k8s.io/cluster-api/api/v1alpha4`. 
Sorry missed this in the review 😄

Arvinderpal commented:
> @Arvinderpal Yeah..I think that's it. We recently bumped to using v1alpha4 types everywhere instead of v1alpha3. Please change all the `sigs.k8s.io/cluster-api/api/v1alpha3` imports to `sigs.k8s.io/cluster-api/api/v1alpha4`.

Thanks for pointing that out. Just made the change. :crossed_fingers:

wfernandes commented:
/lgtm

Arvinderpal commented:
@vincepri should we merge this?

vincepri commented:
/approve
